### ISSUE TITLE
Add image processing support for AVIF

### ISSUE THREAD

**Issue by skyfaller** [gohugoio/hugo#7837]
[AVIF](https://en.wikipedia.org/wiki/AV1#AV1_Image_File_Format_(AVIF)) is a new image format (based on the open AV1 video format) that has the potential to produce significantly smaller file sizes than existing formats, although with different tradeoffs. I find demos of the format to be quite compelling: https://jakearchibald.com/2020/avif-has-landed/

I realize that [webp support](https://github.com/gohugoio/hugo/issues/5924) is probably higher priority, since [caniuse.com has webp at 85% support](https://caniuse.com/?search=webp) while [avif is currently at 25%](https://caniuse.com/?search=avif), but I think it is likely to gain support quickly, since Firefox already has a decent implementation (behind a flag) and Apple are a member of the group that created AV1.

UPDATE August 2021: caniuse.com now has avif at 67.25% support. The next version of Firefox will have it enabled by default, it seems.

**Comment by bep** [comment]
Does a native Go decoder/encoder exist?

**Comment by skyfaller** [comment]
I don't think so, all I was able to find in a quick search was what looks like abandonware: https://github.com/Kagami/go-avif

**Comment by moorereason** [comment]
> ... abandonware: https://github.com/Kagami/go-avif

And it's not native Go.  I binds to a library (libaom) with cgo.

**Comment by 190n** [comment]
Are you willing to call a binary program or C (libaom) or Rust (rav1e) library for AVIF encoding? Since those encoders already have significant engineering work put into them, even if someone did write a native Go encoder, I wouldn't expect it to match the existing ones in performance or compression efficiency.

**Comment by LITUATUI** [comment]
Any news?

**Comment by bep** [comment]
Yes.

**Comment by 190n** [comment]
Is this issue awaiting a native Go encoder? Because I'm not sure if that will ever get written.

**Comment by bep** [comment]
No.

**Comment by brmassa** [comment]
what about https://github.com/Kagami/go-avif?

**Comment by 190n** [comment]
That uses libaom, so it's not a native Go encoder.

**Comment by jothi-prasath** [comment]
+1

**Comment by x2q** [comment]
+1

**Comment by Nacioss** [comment]
Is there any eta for getting AVIF support? The global support among browsers is quite 90% (according to CanIUse).

**Comment by jmooring** [comment]
@Nacioss See <https://github.com/gohugoio/hugo/issues/7837#issuecomment-1063793230>

>  it's not very likely unless a top quality native Go decoder/encoder pops up

So, no, there isn't an ETA, because a top quality native Go decoder/encoder hasn't popped up.

**Comment by BrentHuang** [comment]
+1
avif

**Comment by FlorianLoch** [comment]
Because I am under the impression nobody mentioned it so far: reducing the advantages of AVIF down to better compression isn't, IMHO, entirely fair.

A notable advantage of AVIF for me is HDR support (10- 12 bit per channel compared to JPEG's 8 bit per channel). And for galleries rendered with Hugo that will really get interesting. Perhaps we aren't there yet, but this will certainly draw further attention in the (very) near future.

As I am a Golang developer myself I know about the pain that comes from integrating C dependencies... still I would wish for an alternative to not support this "unless a top quality native Go decoder/encoder pops up" (which very likely might be never, a lot of popular formats have never been properly ported to native Go). 😊

**Comment by LITUATUI** [comment]
For photography JPEG XL is superior, where AVIF shines is in highly compressed images for the WEB. I'm already using it on my Hugo blog (batch conversion with ImageMagick).

**Comment by FlorianLoch** [comment]
> For photography JPEG XL is superior, where AVIF shines is in highly compressed images for the WEB

I agree on this, but as only Safari supports JPEG XL right now, its not really worth considering for web projects in general at the moment: https://caniuse.com/jpegxl

> Since introducing wasm as a "thing in Hugo", I'm planning on replacing the current libwebp bindings with wasm, which would make the threshold for adding AVIF (slightly) lower

That could be a promising approach. Perhaps, at some point in the future, this could allow for pluggable components taking care of wrapping C dependencies providing additional format support to Hugo via some simple interface.

> That said, I'm an avid photographer, and I haven't been totally convinced that my processed photos needs this HDR.

I am into that topic too and so far I was sure I wouldn't need/want it because "HDR" basically referred to oddly saturated and rainbow-colored images. So the term has got a negative reputation with "professional" photographers over the last years. But now with tools like Lightroom, PhotoShop, Affinity supporting HDR in the actual sense of an increased dynamic (and not only weird tone-mapping) it becomes a totally different things. Downside: one requires a display that is actually capable of displaying this (much) wider range. And that is why I am not fully sold yet too, because only few devices are capable of properly presenting such images. But that kind of hardware is on the rise and therefore support for it should be too ;)

**Comment by bep** [comment]
>That could be a promising approach. Perhaps, at some point in the future, this could allow for pluggable components taking care of wrapping C dependencies providing additional format support to Hugo via some simple interface.

Yea, well, there has been done a lot of research about a suitable plugin mechanism in Hugo, and portability, execution speed and good security/sandboxing has been ... hard to find Wasm/WASI is the most promising "thing" I have seen so far. We now embed the Katex wasm in the Hugo binary, but that was a proof of concept. I would very much come to a place where we could define some plugin interfaces where you could include these WASI plugins as Hugo Modules. And some `hugo install` commands that took care of the "packaging".

Whether image processing would be a plugin candidate is up for debate. My motivation behind `libwebp` ported to WASM is mostly motivated by getting rid of the CGO build.

**Comment by 8f5bec34** [comment]
Presumably, since no one mentioned it, this is not a 'top quality native Go decoder/encoder': [https://github.com/takara-ai/AVIF-Go-AI](https://github.com/takara-ai/AVIF-Go-AI)

**Comment by 190n** [comment]
yes, "written entirely by AI" is incompatible with "top quality"

**Comment by cpbotha** [comment]
I recently wrote up my adventures getting HDR images (from my iPhone, but Androids make these too) in a web-compatible form on my Hugo blog: https://cpbotha.net/2025/12/25/weekly-head-voices-263-high-dynamic-life/#the-surprising-trickiness-of-hdr-photos-on-this-blog

Dear @bep thank you so very much for Hugo! I moved my blog (currently 25 years old) from wordpress to Hugo in 2019, and have been enjoying it ever since: https://cpbotha.net/2019/03/31/wordpress-to-hugo/

**Comment by jmooring** [comment]
<https://github.com/gen2brain/avif>

- Initial commit:  9 Feb 2024
- Last commit: 21 May 2025
- Stars: 70

> Go encoder/decoder for [AV1 Image File Format (AVIF)](https://en.wikipedia.org/wiki/AVIF) with support for animated AVIF images (decode only).
> 
> Based on [libavif](https://github.com/AOMediaCodec/libavif) and [aom](https://aomedia.googlesource.com/aom/) compiled to [WASM](https://en.wikipedia.org/wiki/WebAssembly) and used with [wazero](https://wazero.io/) runtime (CGo-free).
> 
> The library will first try to use a dynamic/shared library (if installed) via [purego](https://github.com/ebitengine/purego) and will fall back to WASM.
> 
> Build tags
> `nodynamic` - do not use dynamic/shared library (use only WASM)

**Comment by bep** [comment]
My 50 cents about `avif` (and these Go image decoder/encoder wrappers in general):

>with support for animated AVIF images (decode only).
>Build tags
>nodynamic - do not use dynamic/shared library (use only WASM)

... There's always one or more things that blocks us from using it. The first one (no animation encoder) may not be a blocker, but the last one certainly is. Also, not being able to control the Wazero runtime (it's one global runtime) is also  non-go for us. And then there is this trust issue with third-party C libraries; this may be less of a thing with WASM, but it's still a thing.

I added the WASM webp support in Hugo partly as an exercise to see how loosely coupled and format agnostic I could get the integration. I you look at https://github.com/gohugoio/hugo/blob/master/internal/warpc/webp.go and rename the webp types to something generic, there's just some config maps that differers from webp and some other format. It would be cool if a WASM file could be mounted in `hugo.toml` with that config map passed in. You would still have a trust issue re the WASM origin, but 1. It's a sandboxed runtime and 2. The dependency would live outside of Hugo, so up to you.

**Comment by bep** [comment]
OK, starting looking at AVIF HDR images today (e.g. https://raw.githubusercontent.com/AOMediaCodec/av1-avif/master/testFiles/Netflix/avif/hdr_cosmos01650_cicp9-16-9_yuv420_limited_qp10.avif) and it convinced me.

**Comment by bep** [comment]
Some notes to self:  JPEG also support HDR output, not sure how Go's decoder handles it (will test). But exporting some firework photos from last night in 2048 px wide in both, ticks JPEG in at 1.3 MiB and AVIF at 500 KiB, which makes the latter much more practical.

**Comment by aariyantoyeb-hash** [comment]
Good luck
